### PR TITLE
Add speed profiler

### PR DESCRIPTION
**Pull request by officialgr**
## Summary
Adds a report-only speed profiling workflow for Lucebox inference changes.

## Changes
- Adds `server/scripts/profile.py`
- Adds `.github/workflows/speed-profile.yml`
- Adds profiler documentation at `server/docs/speed-profile.md`

## Notes
- Workflow is report-only / non-blocking.
- Intended for the self-hosted RTX 3090 GPU runner.
- Produces `profile.json`, `profile.md`, and nsys artifacts when available.

<a href="https://cubic.dev/pr/Luce-Org/lucebox-hub/pull/431?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

**Comment by davide221** [comment]
@officialgr thanks for the profiler work, can you fix the CI error?

**Comment by officialgr** [comment]
@davide221 Fixed — the job was running profile.py with the system python3, which doesn't have transformers (used for tokenization). It now installs it into an isolated venv before running.

**Comment by officialgr** [comment]
The failure was caused by the speed-profile workflow looking for the wrong speculative draft model path. The workflow now defaults to the repo-documented Qwen3.6 draft GGUF path and skips cleanly with diagnostics if runner-local weights are missing.

**Comment by davide221** [comment]
Nice progress, it now runs end-to-end on the 3090 and is already useful. Two things to look at in your own passing run before this is mergeable:

1. The lossless gate reports FAIL (2/3 prompts diverge) but the job is still green. Is that a real correctness bug or too-strict a check?
2. The nsys section prints 0.0 everywhere with an empty table

A couple of things would make this much more valuable: a baseline + threshold so it can actually flag a regression and a thought on whether we should also profile dflash_server (the shipping path), since test_dflash can't see the prefix cache / batching.

**Comment by officialgr** [comment]
@davide221 
Ran it end-to-end on the 3090 at the shipping defaults (Qwen3.6-27B, budget 22, n_gen 48).

**1. Lossless — gate is no longer "too strict."** Added an AR-vs-AR determinism control (reuses the existing AR passes, no extra GPU cost): a prompt only FAILs when AR is self-deterministic *and* spec-decode still differs; intrinsically-nondeterministic prompts become *inconclusive*. On my run it flags a divergence at token #44 (2/3 prompts). I checked whether it's a tail artifact by re-running at n_gen 96 — it stayed at #44, so it's reproducible/structural, not a length effect. I'm deliberately not over-calling it: it's ruled out as run-to-run noise, but it could still be batched-verify FP (verify scores draft tokens as a batch vs AR one-at-a-time). The definitive bug-vs-FP test is the logit gap at #44 (near-tie = FP, clear gap = bug) — the binaries don't emit per-token logits yet, so I'm flagging it for the engine team rather than claiming a bug.

**2. nsys — fixed, now fully populated.** Root cause was report-name drift across nsys versions (`cuda_gpu_kern_sum` ↔ legacy `gpukernsum`) plus a swallowed error. It now tries both name sets and, if stats still won't parse, prints the nsys error + version instead of silent zeros. On Qwen3.6: 568.8 ms GPU kernel time, 674.6 launches/token, top kernel `mul_mat_q` (187 ms / 2880 launches), 75.7 ms/token in host↔device copies.

**3. Baseline + threshold — added.** `--baseline` + `--regress-pct` (default 10%, wide enough for run-to-run variance), direction-aware (tok/s & AL up = good; ms/token & TTFT down = good). The report flags `⚠️ REGRESSION` and CI emits a non-blocking warning. Verified the delta path on a self-vs-self run (`✅ within threshold`). Seeded `server/scripts/speed-baseline.json` from the green run.

**4. dflash_server — recommendation (no code, since you asked for "a thought").** Keep the binaries as the per-PR signal (deterministic, low-noise, isolates compute). The shipping path's prefix-cache + batching is a different question (throughput under load), so I'd add it as a separate, higher-variance Phase 2 — replay a request trace at fixed concurrency; TTFT/p50/p99 + cache-hit rate — on a schedule/label, not every PR. Written up in `docs/specs/speed-profile.md`.

**Final note on `n_gen`.** `n_gen=48` is good for CI speed, but it is too short to give the most representative throughput signal — in my runs it stays around ~0.98 and is more sensitive to setup/early-step overhead. The `128–256` range works much better and gives more stable speed numbers. I’d recommend moving the default to `128` for a balanced PR profile, or `256` if we want the profile to favor benchmark quality over runtime.

**Comment by dusterbloom** [comment]
@officialgr  Great work getting the profiler reliable and report-only. 

1 thing I’d love to see before we add more features: let’s make the profiler honest about its own noise. Right now a single run produces one latency number, which makes it hard to tell whether a 3% delta in a PR is real or just run-to-run jitter.

Here a couple of things that could help avoid chasing ghost / thermal variance that could be added:

   • More than 3 repeated runs per prompt (e.g., 3–5, configurable) and report the mean + stddev.
   • A clear “noisy / below detection threshold” state when the relative stddev is above some sensible threshold, or when the before/after overlap is too large to call a regression.
   • Keep it report-only for now - warnings, not blockers.
   
Keep up the good work. Let me know what you think.

**Comment by officialgr** [comment]
@dusterbloom 
Thanks — agreed with this direction, and I think the profiler now covers the noise concerns while staying report-only.

What changed:

- The profiler now uses a longer generation sample by default: `--n-gen 128`. This controls the number of generated response tokens per benchmark prompt. Based on testing, shorter generations below 128 were too sensitive to startup/thermal jitter, while 128 is a good compromise between CI runtime and measurement quality.
- The profiler now runs repeated timing passes per prompt by default (`--reps 5`), with `--reps` still configurable for smoke runs or future tuning.
- Instead of reporting one latency/throughput number, it records per-run samples and reports mean ± stddev for the headline metrics, including decode tok/s, ms/token, TTFT estimate, and acceptance length.
- The markdown report now has a dedicated “Noise / detection threshold” section. If the relative stddev exceeds the configured threshold (`--noise-rsd-pct`, default 5%), it marks the profile as NOISY and explicitly says to treat small deltas as below the profiler detection threshold.
- Baseline comparisons now include stddev-aware overlap handling. If the current and baseline 1σ ranges overlap and the delta is smaller than the regression threshold, the row is marked as “noisy / overlap” rather than encouraging us to chase a likely ghost regression.
- This is still report-only. The workflow only emits GitHub warning annotations for noisy profiles / regressions / losslessness issues; it does not block the PR.
- CI was updated to use the same defaults: `--n-gen 128`, `--reps 5`, and `--noise-rsd-pct 0.05`, with repo-variable overrides available for temporary smoke runs.

So the profiler should now be more honest about whether a small delta is meaningful or just run-to-run / thermal variance. The `--n-gen 128` change helps reduce startup-noise sensitivity, the repeated runs quantify remaining jitter, and the noisy/overlap labels keep the whole thing report-only instead of turning measurement noise into a blocker.

File: docs/specs/speed-profile.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
+# Lucebox speed profiler (MVP)
+
+A small, CI-runnable profiler for the inference engine. It measures forward-pass
+speed on one GPU and produces a report that shows **where the time goes**, so a
+reviewer can see at a glance whether a PR moved the needle and where the next
+optimization margin is.
+
+It runs the engine **binaries directly** (no HTTP) so the numbers reflect compute,
```
Comment (howard0su) [inline docs/specs/speed-profile.md]: which binary?

File: docs/specs/speed-profile.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+## CI settings
+
+The `Speed Profile` workflow uses the same profiler defaults as the local recipe:
+`--n-gen 128`, `--reps 5`, and `--noise-rsd-pct 0.05`. Runner owners can
+temporarily override those values with repo variables `LUCEBOX_SPEED_N_GEN`,
+`LUCEBOX_SPEED_REPS`, and `LUCEBOX_SPEED_NOISE_RSD_PCT`, but PR-to-PR comparisons
+should keep them fixed.
+
+## Run it locally
```
Comment (howard0su) [inline docs/specs/speed-profile.md:101]: how to compare the result with baseline?

File: server/scripts/profile.py [hunk 3]
```diff
@@ -0,0 +1,822 @@
...
+              "def sum_product(numbers: List[int]) -> Tuple[int, int]:\n"
+              '    """Return a tuple of (sum, product) of all integers. Empty -> (0, 1).\n'
+              "    >>> sum_product([1, 2, 3, 4])\n    (10, 24)\n"
+              '    """\n    s = 0\n    p = 1\n    for n in')},
+]
+
+# --------------------------------------------------------------------------------------
+# stdout parsers (matched to the exact print formats in test_dflash.cpp / test_generate.cpp)
+# --------------------------------------------------------------------------------------
+RE_PREFILL      = re.compile(r"\[prefill\]\s+(\d+)\s+tokens?\s+in\s+([\d.]+)\s*s")
```
Comment (howard0su) [inline server/scripts/profile.py:73]: this is error prone. it is fine for the first version. I would suggest to add json output support to test_dflash binary so we don't do the text parse here.

**Comment by officialgr** [comment]
Thanks for the review, @howard0su — good calls, addressed all three:

1. Which binary: there are two — `test_dflash` (the speculative / DFlash decode
path) and `test_generate` (the autoregressive baseline), both built from server/
as CMake targets (`build/test_dflash`, `build/test_generate`, overridable via
`--df-bin`/`--ar-bin` or `DFLASH_BIN`/`DFLASH_BIN_AR`). I've named them explicitly
in the doc.

2. Comparing against a baseline: save a run's JSON (`--out-json
scripts/speed-baseline.json`) and commit it, then on later runs pass `--baseline
scripts/speed-baseline.json --regress-pct 0.10`. The report renders a "Delta vs
baseline" table (baseline ± σ, now ± σ, Δ, Δ%) and flags a row as a regression
only when the move exceeds `--regress-pct` and the 1σ intervals don't overlap.
Added a "Comparing against a baseline" section to the doc spelling out the
round-trip.

3. Text parsing: agreed, it's coupled to the exact print formats in
test_dflash.cpp / test_generate.cpp and is brittle. Keeping it for this first
version as you suggested; I'll open a follow-up to add a `--json` output mode to
the binaries and switch the profiler to consume that (keeping the text path as a
fallback for one release).
